packages/platform-browser/src/dom/shared_styles_host.ts

```diff
@@ -20,94 +20,159 @@ import {
 /** The style elements attribute name used to set value of `APP_ID` token. */
 const APP_ID_ATTRIBUTE_NAME = 'ng-app-id';
 
+/**
+ * A record of usage for a specific style including all elements added to the DOM
+ * that contain a given style.
+ */
+interface UsageRecord<T> {
+  elements: T[];
+  usage: number;
+}
+
+/**
+ * Removes all provided elements from the document.
+ * @param elements An array of HTML Elements.
+ */
+function removeAll(elements: Iterable<HTMLElement>): void {
+  for (const element of elements) {
+    element.remove();
+  }
+}
+
 @Injectable()
 export class SharedStylesHost implements OnDestroy {
-  // Maps all registered host nodes to a list of style nodes that have been added to the host node.
-  private readonly styleRef = new Map<
-    string /** Style string */,
-    {
-      elements: HTMLStyleElement[];
-      usage: number;
-    }
-  >();
-  private readonly hostNodes = new Set<Node>();
-  private readonly styleNodesInDOM: Map<string, HTMLStyleElement> | null;
-  private readonly platformIsServer: boolean;
+  /**
+   * Provides usage information for embedded style content and the HTML elements containing each style.
+   */
+  private readonly embeddedStyles = new Map<string /** content */, UsageRecord<HTMLStyleElement>>();
+
+  /**
+   * Set of host DOM nodes that will have styles attached.
+   */
+  private readonly hosts = new Set<Node>();
+
+  /**
+   * A lookup for server rendered styles if any are present in the DOM on initialization.
+   * `null` if no server rendered styles are present in the DOM.
+   */
+  private readonly serverStyles: Map<string, HTMLStyleElement> | null;
+
+  /**
+   * Whether the application code is currently executing on a server.
+   */
+  private readonly isServer: boolean;
 
   constructor(
     @Inject(DOCUMENT) private readonly doc: Document,
     @Inject(APP_ID) private readonly appId: string,
-    @Inject(CSP_NONCE) @Optional() private nonce?: string | null,
-    @Inject(PLATFORM_ID) readonly platformId: object = {},
+    @Inject(CSP_NONCE) @Optional() private readonly nonce?: string | null,
+    @Inject(PLATFORM_ID) platformId: object = {},
   ) {
-    this.styleNodesInDOM = this.collectServerRenderedStyles();
-    this.platformIsServer = isPlatformServer(platformId);
-    this.resetHostNodes();
+    this.isServer = isPlatformServer(platformId);
+    this.serverStyles = this.findServerStyles();
+    this.hosts.add(this.doc.head);
```

Comment:
nit: we can probably drop `this.`:
```suggestion
    this.hosts.add(doc.head);
```